src/server.py

```diff
@@ -17,38 +18,101 @@
     except ValidationError:
         raise HTTPException(500, "Invalid configuration files")
 
+    return {
+        city_id: CachedCityDates(
+            city_configuration=city_config,
+            available_dates=city_data_cache.get_cached_dates(city_id),
+        )
+        for city_id, city_config in cities_config.items()
+    }
+
 
 @app.get("/cities/{city_id}")
-def get_city_data(city_id: str, weekday: str | None = None) -> ResponseCityData:
-    try:
-        weekday_enum: Weekday = Weekday.get_by_value_with_default(weekday)
-    except ValueError as exc:
-        raise HTTPException(400, str(exc))
+def get_city_data(
+    city_id: str,
+    weekday: Weekday | None = Depends(validate_weekday),
+    date: datetime.date | None = Depends(validate_date),
+) -> ResponseCityData:
+    """
+    Returns processed data for a given city.
+
+    - Without parameters: returns data for the current day
+    (from cache if available, otherwise built and cached).
+    - With `weekday`: builds and returns data for the given day of the week.
+    - With `date`: returns cached data for the given date if available,
+    otherwise responds with 404.
+    - With both `weekday` and `date`: responds with 400, only one parameter is allowed.
+    """
+    if date is not None and weekday is not None:
+        raise HTTPException(400, "Provide either date or weekday")
+
+    if date:
+        if cached_data := city_data_cache.get(city_id, date):
+            return cached_data
+
+        raise HTTPException(
+            404, f"City data for {city_id} not found in cache for {date}"
+        )
+
+    is_today_build = weekday is None
+    if is_today_build:
+        date = datetime.date.today()
+        weekday = Weekday.get_current()
+
+        if today_cached_data := city_data_cache.get(city_id, date):
+            return today_cached_data
 
     if (city_configuration := CityConfiguration.get_by_city_id(city_id)) is None:
-        raise HTTPException(404, "City not found")
-
-    if not city_data_cache.is_fresh(city_id, weekday_enum):
-        try:
-            city_data_builder = CityDataBuilder(city_configuration, weekday_enum)
-            city_data_cache.store(city_id, weekday_enum, city_data_builder)
-        except Exception as exc:
-            logger.exception(
-                f"Failed to build city data for city {city_id} and weekday {weekday_enum}",
-                exc_info=exc,
-            )
-
-    if city_data := city_data_cache.get(city_id, weekday_enum):
-        return city_data
-
-    raise HTTPException(500, f"Data processing for {city_id} failed")
+        raise HTTPException(404, f"City {city_id} not found")
+
+    assert weekday is not None
+    assert date is not None
+
+    try:
+        city_data_builder = CityDataBuilder(city_configuration, weekday)
+    except Exception as exc:
+        logger.exception(
+            f"Failed to build city data for city {city_id} for weekday {weekday}",
+            exc_info=exc,
```

Comment:
The recommended approach is to use lazy formatting for log messages (due to formatting being expensive so we want to do it only if we actually want to display this log message).
```suggestion
            "Failed to build city data for city %s for weekday %s",
            city_id,
            weekday
            exc_info=exc,
```